testing/tests/audio.lua

```diff
@@ -0,0 +1,446 @@
+-- love.audio
+
+
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+------------------------------------OBJECTS-------------------------------------
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+
+
+-- RecordingDevice (love.audio.getRecordingDevices)
+love.test.audio.RecordingDevice = function(test)
+  if GITHUB_RUNNER == true then
+    return test:skipTest('cant emulate recording devices in CI')
+  end
+  -- check devices first
+  local devices = love.audio.getRecordingDevices()
+  if #devices == 0 then
+    return test:skipTest('cant test this works: no recording devices found')
+  end
+  -- check object created and basics
+  local device = devices[1]
+  test:assertObject(device)
+  test:assertMatch({1, 2}, device:getChannelCount(), 'check channel count is 1 or 2')
+  test:assertNotEquals(nil, device:getName(), 'check has name')
+  -- check initial data is empty as we haven't recorded anything yet 
+  test:assertNotNil(device:getBitDepth())
+  test:assertEquals(nil, device:getData(), 'check initial data empty')
+  test:assertEquals(0, device:getSampleCount(), 'check initial sample empty')
+  test:assertNotNil(device:getSampleRate())
+  test:assertEquals(false, device:isRecording(), 'check not recording')
+  -- start recording for a short time
+  -- @TODO needs delay for VMs
+  local startrecording = device:start(32000, 4000, 16, 1)
+  test:waitFrames(10)
+  test:assertEquals(true, startrecording, 'check recording started')
+  test:assertEquals(true, device:isRecording(), 'check now recording')
+  test:assertEquals(4000, device:getSampleRate(), 'check sample rate set')
+  test:assertEquals(16, device:getBitDepth(), 'check bit depth set')
+  test:assertEquals(1, device:getChannelCount(), 'check channel count set')
+  local recording = device:stop()
+  test:waitFrames(10)
+  -- after recording 
+  test:assertEquals(false, device:isRecording(), 'check not recording')
+  test:assertEquals(nil, device:getData(), 'using stop should clear buffer')
+  test:assertObject(recording)
+end
+
+
+-- Source (love.audio.newSource)
+love.test.audio.Source = function(test)
+  -- create stereo source
+  local stereo = love.audio.newSource('resources/click.ogg', 'static')
+  test:assertObject(stereo)
+  -- check stereo props
+  test:assertEquals(2, stereo:getChannelCount(), 'check stereo src')
+  test:assertEquals(66, math.floor(stereo:getDuration("seconds")*1000), 'check stereo seconds')
+  test:assertNotNil(stereo:getFreeBufferCount())
+  test:assertEquals('static', stereo:getType(), 'check stereo type')
+  -- check cloning a stereo
```

Comment:
Can each section like this have an empty line before it? Things are super compressed in the test code right now, which makes it hard to read